src/kernels/MIOpenDropoutHIP.cpp

```diff
@@ -0,0 +1,374 @@
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ *******************************************************************************/
+#ifndef MIOPEN_DONT_USE_HIP_RUNTIME_HEADERS
+#include <hip/hip_fp16.h>
+#include <hip/hip_runtime.h>
+#endif
+
+// Workaround to overcome redefinition errors while including rocrand header files directly
+#include "miopen_rocrand.hpp"
+
+#ifndef MIOPEN_USE_FP32
+#define MIOPEN_USE_FP32 0
+#endif
+
+#ifndef MIOPEN_USE_FP16
+#define MIOPEN_USE_FP16 0
+#endif
+
+#if MIOPEN_USE_FP16 == 1
+#define _FLOAT half
+#endif
+#if MIOPEN_USE_FP32 == 1
+#define _FLOAT float
+#endif
+
+#ifndef RUN_FORWARD
+#define RUN_FORWARD 0
+#endif
+
+#ifndef RUN_INIT_PRNG
+#define RUN_INIT_PRNG 0
+#endif
+
+#if RUN_INIT_PRNG // Initialize PRNG
+/**
+ * @brief Initializes the kernel state.
+ *
+ * This function initializes the kernel state by assigning a random state to each element in the
+ * state array.
+ *
+ * @param state Pointer to the array of rocrand_state_xorwow structures representing the kernel
+ * state.
+ * @param prng_seed The seed value for the pseudo-random number generator.
+ * @param states_num The number of elements in the state array.
+ */
+extern "C" __global__ void
+InitKernelStateHIP(rocrand_state_xorwow* state, ulong prng_seed, ulong states_num)
+{
+    // Get the index of the current element
+    size_t index  = blockIdx.x * blockDim.x + threadIdx.x;
+    size_t stride = blockDim.x * gridDim.x;
+
+    for(size_t gid = index; gid < states_num; gid += stride)
+    {
+        rocrand_state_xorwow state_gid;
+        rocrand_init(prng_seed, gid, 0ULL, &state_gid);
+        state[gid] = state_gid;
+    }
+}
+#endif
+
+#if !RUN_INIT_PRNG
+
+#if RUN_FORWARD
+#ifndef USE_MASK
+#define USE_MASK 0
+#endif
+
+#ifndef USE_RSVSP
+#define USE_RSVSP 0
+#endif
+
+// Dropout Forward
+/**
+ * @brief Data is loaded into registers from the input.
+ * The mask is either generated and saved in the reserved space, or an existing mask is loaded into
+ * registers from the reserved space. The loaded mask is applied to the data in the registers. The
+ * processed data is then written to the output. Reading and writing of input data can be performed
+ * as vectorized operations, with the extent of vectorization determined by `#RD_BLOCK`.
+ */
+template <typename F, typename T, typename B, bool MASK = false, bool RSVSP = false>
+__forceinline__ __device__ void dropoutfw(const rocrand_state_xorwow* state,
+                                          float dropout,
+                                          float scale,
+                                          int dim1,
```

Comment:
`dims` numbering should also start at `0` for consistency.

Reply:
You can also pass `dims` and `strides` as `int dim[N]` where `N` is a template parameter. Then we can handle `ncdhw`, i.e. 3D spatial tensors as well.

Reply:
I think the dims are numbered correctly, it is a flattened 5D tensor.